src/Radicle/Internal/ConcurrentMap.hs

```diff
@@ -14,67 +13,61 @@ import           Protolude hiding (empty)
 
-fromMap :: Map k v -> IO (CMap k v)
-fromMap m = CMap <$> (traverse newMVar m >>= newMVar)
-
 -- | Atomically lookup a value.
 lookup :: Ord k => k -> CMap k v -> IO (Maybe v)
 lookup k (CMap m_) = withMVar m_ $ \m -> do
    case Map.lookup k m of
      Nothing -> pure Nothing
-     Just v_ -> pure <$> readMVar v_
+     Just v_ -> readMVar v_
 
 -- | Non-atmoically read the contents of a 'CMap'. Provides a
 -- consistent shapshot of which keys were present at some
 -- time. However, the values might be snapshotted at different times.
 nonAtomicRead :: CMap k v -> IO (Map k v)
 nonAtomicRead (CMap m_) = do
   m <- readMVar m_
-  traverse readMVar m
+  Map.mapMaybe identity <$> traverse readMVar m
 
 -- | Atomically insert a key-value pair into a 'CMap'.
 insert :: Ord k => k -> v -> CMap k v -> IO ()
 insert k v (CMap m_) = modifyMVar m_ $ \m -> do
-  v_ <- newMVar v
+  v_ <- newMVar (Just v)
   let m' = Map.insert k v_ m
   pure (m', ())
 
 -- | Atomically modifies a value associated to a key but only if it
--- exists. There is no guarantee the value is still associated with
--- the same key, or any key, by the time the operation completes.
+-- exists. Blocks all operations on @k@ while @f@ is running.
 modifyExistingValue :: (MonadUnliftIO m, Ord k) => k -> CMap k v -> (v -> m (v, a)) -> m (Maybe a)
-modifyExistingValue k (CMap m_) f = withRunInIO $ \runInIO -> do
-  m <- liftIO $ readMVar m_
-  case Map.lookup k m of
-    Nothing -> pure Nothing
-    Just v_ -> do
-      x <- modifyMVar v_ (runInIO . f)
-      pure (Just x)
-
--- | Atomically modifies the value associated to a key.
-modifyValue :: (MonadUnliftIO m, Ord k) => k -> CMap k v -> (Maybe v -> m (Maybe v, a)) -> m a
-modifyValue k (CMap m_) f = withRunInIO $ \runInIO -> modifyMVar m_ $ \m ->
-  case Map.lookup k m of
-    Nothing -> do
-      (maybev, x) <- runInIO $ f Nothing
-      case maybev of
+modifyExistingValue k c f =
+    modifyValue k c $ \case
+        Nothing -> pure (Nothing, Nothing)
         Just v -> do
-          v_ <- newMVar v
-          pure (Map.insert k v_ m, x)
-        Nothing -> pure (m, x)
-    Just v_ -> modifyMVar v_ $ \v -> do
-      (maybev', x) <- runInIO $ f (Just v)
-      case maybev' of
-        Just v' -> pure (v', (m, x))
-        Nothing -> pure (v, (Map.delete k m, x))
+            (v', a) <- f v
+            pure (Just v', Just a)
+
+-- | Atomically modifies a value associated to a key. Blocks all
+-- operations on @k@ while @f@ is running.
+modifyValue :: forall k v a m. (MonadUnliftIO m, Ord k) => k -> CMap k v -> (Maybe v -> m (Maybe v, a)) -> m a
+modifyValue k (CMap m_) f = withRunInIO $ \runInIO -> do
+    valueVar <- modifyMVar m_ $ \m ->
+        case Map.lookup k m of
+            Nothing -> do
+                valueVar <- newMVar Nothing
+                pure (Map.insert k valueVar m, valueVar)
+            Just valueVar -> do
+                pure (m, valueVar)
+    -- To prevent space leaks we probably should remove the key from
```

Comment:
With this new version I think the only way to get rid of the `Nothing`s would be to schedule a cleanup. We don't have a "delete machine" endpoint so it's really not a problem for us (yet!).